### PR TITLE
Fix #238 - Enhance Slot to produce values for all possible types

### LINKED ISSUE #238: Enhance Slot to produce values for all possible types

### PR DESCRIPTION
**Pull request by gumbelmj**
This enhances Slot to allow it to produce a decimal value, detect NotAvailable and Error in the values and print the value with the units of measure.

### CODE REVIEW

**Review by mmschaefer — APPROVED** [review]

File: src/org/etools/j1939_84/bus/j1939/packets/Slot.java [hunk 1]
```diff
@@ -19,197 +18,318 @@
...
+    }
+
+    /**
+     * Returns a scaled value. That is result = value * scaling + offset
+     *
+     * @param value the value to scale
+     * @return double
+     */
+    public double scale(double value) {
+        if (getScaling() != null && getOffset() != null) {
```
Comment (battjt) [inline src/org/etools/j1939_84/bus/j1939/packets/Slot.java:204]: Should probably be
double v = value;
if (getScaling()!=null)v*=getScaling();
if(getOffset()!=null)v+=getOffset();
return v;

File: src/org/etools/j1939_84/bus/j1939/packets/Slot.java [hunk 2]
```diff
@@ -19,197 +18,318 @@
...
+
+    private long flipBytes(byte[] data, int byteLength) {
+        long value = 0;
+        for (int i = 0; i < byteLength; i++) {
+            value += ((long) (data[i] & 0xFF)) << i * 8;
+        }
+        return value;
+    }
+
+    private long mask() {
```
Comment (battjt) [inline src/org/etools/j1939_84/bus/j1939/packets/Slot.java:307]: long mask = ~0L >>> (64 - len);

File: src/org/etools/j1939_84/bus/j1939/packets/Slot.java [hunk 3]
```diff
@@ -19,197 +18,318 @@
...
+                final String type = values[2];
+                final Double scaling = doubleValue(values[3]);
+                final String unit = values[4];
+                final Double offset = doubleValue(values[5]);
+                final int length = intValue(values[6], -1);
+                Slot slot = new Slot(id, name, type, scaling, offset, unit, length);
+                slots.put(id, slot);
+            }
+        } catch (Exception e) {
+            J1939_84.getLogger().log(Level.SEVERE, "Error loading map from slots", e);
```
Comment (battjt) [inline src/org/etools/j1939_84/bus/j1939/packets/Slot.java:105]: I think a failure to load the file should be a runtime exception, not just a log message.

Comment (gumbelmj) [reply src/org/etools/j1939_84/bus/j1939/packets/Slot.java:105]: Runtime exception would cause the app to crash.

Comment (battjt) [reply src/org/etools/j1939_84/bus/j1939/packets/Slot.java:105]: If we don't crash the app, then the output may be wrong.  The user should be notified that the results will be invalid.

**Pull request merged**